### ISSUE TITLE
Thumbnails loading takes longer as you reach the bottom of larger playlist

### ISSUE THREAD

**Issue by Zal0** [libretro/RetroArch#10676]
# First and foremost consider this:
- Only RetroArch bugs should be filed here. Not core bugs or game bugs
- This is not a forum or a help section, this is strictly developer oriented

## Description

The first entries of a playlist load the thumbnails pretty fast but as you go down in the list the time to load a thumbnail gets longer and longer

This is specially visible in ozone with 2000 entries in a playlist. It also happens in rgui and glui but it is less annoying since these two menu drivers wait a few miliseconds before loading the thumbnail (this option is not available in ozone)

### Expected behavior

Loading the thumbnail of the first entry on the list and on the last one should take the same time

### Actual behavior

Loading thumbnails at the end of a big playlist takes a lot of time compared to the first entries

### Steps to reproduce the bug

1. Create a playlist with 2000 entries  (the full romset snes playlist)
2. Configure thumbnails
3. Start navigating the playlist, thumbnails loading times are small
4. Go to letter Z and compare

### Bisect Results

Not sure when this started to happen, but it's been there in the recent releases

### Version/Commit

- RetroArch: 1.8.7 stable

### Environment information

- OS: I am testing this on the Nintendo Switch, but it's very likely that it's happening in all platforms

**Comment by i30817** [comment]
This is not surprising if the loading tasks are a unlimited FIFO queue (instead of a limited or unlimited LIFO queue).

**Comment by jdgleaver** [comment]
Yes, we're planning to unify the thumbnail handling across all menu drivers, so XMB/Ozone get the same benefits as Material UI with regards to loading efficiency.

No ETA yet, though (currently being crushed under the weight of my immediate TODO list...)

**Comment by Zal0** [comment]
But any idea why loading the last thumbnail in the list takes like 10 times compared to the first one?

**Comment by i30817** [comment]
It's a simple. Retroarch has a queue of 'tasks' it needs done in another thread.

You start empty: [] ; add one at the end: [A]; and others at the end [A,B,C,D]; etc

Then the thread comes along as goes 'oh to be fair, the first tasks should go first' (this is wrong in this case but it's right in general), and processes A; the tasks list is [B,C,D]. What's happening is that when you scroll down you're adding tasks faster than the thread can execute them, so what the thread has to do until it reaches the 'end' task which is the last thumbnail you requested is gettiing longer and longer.

You could avoid this by either making the thread process the last added (LIFO) instead of the first added (FIFO), or just by only adding the tasks you really need, which is what the delay does; only adds the task to decode the png if the user stopped on a game entry for 'long enough' which scrolling doesn't.

**Comment by Zal0** [comment]
I understand what you are saying but not sure this is the problem...

If instead of starting on letter A I press Up and go to the last title no queued tasks could have been added (unless it's adding the whole list of thumbnails for some unknown reason). Then if I wait until the thumbnail is loaded and press up again I can clearly see that it takes around 2 seconds to load the thumbnail. 

Also, in Ozone that queue doesn't seem to exists since the menu stays frozen until the thumbnail is loaded, meaning I cannot create new tasks until the previous one has finished

**Comment by i30817** [comment]
There is a option in retroarch  to disable/enable 'threaded tasks' in the advanced options (after enabling show advanced options) in the 'Settings->User Interface' submenu. Check if it's turned on.

You description sounds like you turned it off imo.

**Comment by jdgleaver** [comment]
This is incredibly strange...

i30817 describes the situation perfectly. And it does kinda sound like you have threading disabled...

> If instead of starting on letter A I press Up and go to the last title no queued tasks could have been added (unless it's adding the whole list of thumbnails for some unknown reason). Then if I wait until the thumbnail is loaded and press up again I can clearly see that it takes around 2 seconds to load the thumbnail.

...But regardless of threading settings, I can't reproduce that on any of my systems. Navigating from the last entry to the first is identical (code and performance-wise) to going from the first to the second.

> Also, in Ozone that queue doesn't seem to exists since the menu stays frozen until the thumbnail is loaded, meaning I cannot create new tasks until the previous one has finished

That's definitely not normal - even with threading disabled, it loads the PNG image in chunks, which may reduce frame rate but wouldn't hang the menu entirely.

It seems there's something more fundamentally wrong with your setup. I'm not sure what that could be, though - I've never seen this kind of issue before.

**Comment by Zal0** [comment]
Threaded tasks are enabled (although I see no difference) and On-Demand thumbnails Downloads is off
I have deleted my full retroarch.cfg and recreated the playlist from scratch... still happening

Could this be just a bug on the Nintendo Switch?
Do thumbnails compression or resolution affect any of this?
This only happens with very big playlists (around 2000 entries), and it is more noticeable if you load 2 thumbnails (boxart and screenshot)

**Comment by jdgleaver** [comment]
> Do thumbnails compression or resolution affect any of this?

Larger thumbnails will take longer to load (so box arts are the worst offenders here), but I've not found any images in the thumbnails repo that cause noticeable delays

> This only happens with very big playlists (around 2000 entries), and it is more noticeable if you load 2 thumbnails (boxart and screenshot)

> Could this be just a bug on the Nintendo Switch?

I can only think this might be a Switch issue... I've tested thumbnail loading on a host of platforms, all the way down to the lowly o3DS, and I've never seen the behaviour you're describing. Unfortunately, I don't own a Switch, so I'll not be much use in debugging this :(

**Comment by Zal0** [comment]
Somo feedback about this issue still present on latest build
With the new delay introduced before loading the thumnails things have improved a bit but still there is something weird here.
- If I disable the **second thumbnail** then everything goes smooth even on the biggest playlists like Mame (around 4500 entries). I think this has been fixed on recent builds
- Enabling the second thumbnail makes the issue to happen again
- If I enable the **second thumbnail** but then press Y to change it until it disappears  to show the rom stats the issue is still happening even thought it's not supposed to be loading anything
- If I Disable the first thumbnail but enable the second then the issue disappears so it doesn't seem to be related to just the second thumbnail. This happens when there are two thumbnails enabled
